Implement relaxed SIMD madd and nmadd instructions

danlliu opened the pull request:
#### 5236c954e0b8cc779c14498d7530ff8d59b7d0a0
<pre>
Implement relaxed SIMD madd and nmadd instructions
<a href="https://bugs.webkit.org/show_bug.cgi?id=258047">https://bugs.webkit.org/show_bug.cgi?id=258047</a>
rdar://110736062

Reviewed by Justin Michaud.

Added implementation and tests for f32x4.relaxed_madd, f32x4.relaxed_nmadd, f64x2.relaxed_madd, f64x2.relaxed_nmadd.

* JSTests/wasm/stress/simd-const-relaxed-f32-madd.js: Added.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.export.string_appeared_here.param.sz.i32.result.f32.i32.const.9.drop.v128.const.f32x4.5.0.4.0.37.0.6.0.v128.const.f32x4.22.0.25.0.3.0.20.0.v128.const.f32x4.1.0.1.0.0.0.1.0.f32x4.relaxed_madd.return.f32x4.extract_lane.1.func.export.string_appeared_here.param.sz.i32.result.f32.i32.const.9.drop.v128.const.f32x4.5.0.4.0.37.0.6.0.v128.const.f32x4.22.0.25.0.3.0.20.0.v128.const.f32x4.1.0.1.0.0.0.1.0.f32x4.relaxed_nmadd.return.f32x4.extract_lane.3.async test):
* JSTests/wasm/stress/simd-const-relaxed-f64-madd.js: Added.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.export.string_appeared_here.param.sz.i32.result.f64.i32.const.9.drop.v128.const.f64x2.28.0.24.0.v128.const.f64x2.10.0.20.0.v128.const.f64x2.1.0.2.0.f64x2.relaxed_madd.return.f64x2.extract_lane.0.func.export.string_appeared_here.param.sz.i32.result.f64.i32.const.9.drop.v128.const.f64x2.28.0.24.0.v128.const.f64x2.10.0.20.0.v128.const.f64x2.1.0.2.0.f64x2.relaxed_nmadd.return.f64x2.extract_lane.1.async test):
* Source/JavaScriptCore/assembler/ARM64Assembler.h:
* Source/JavaScriptCore/assembler/MacroAssemblerARM64.h:
(JSC::MacroAssemblerARM64::vectorFusedMulAdd):
(JSC::MacroAssemblerARM64::vectorFusedNegMulAdd):
* Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
(JSC::MacroAssemblerX86_64::vectorMulAdd):
* Source/JavaScriptCore/b3/B3LowerToAir.cpp:
* Source/JavaScriptCore/b3/B3Opcode.cpp:
(WTF::printInternal):
* Source/JavaScriptCore/b3/B3Opcode.h:
* Source/JavaScriptCore/b3/B3SIMDValue.h:
* Source/JavaScriptCore/b3/B3Validate.cpp:
* Source/JavaScriptCore/b3/B3Value.cpp:
(JSC::B3::Value::effects const):
(JSC::B3::Value::key const):
* Source/JavaScriptCore/b3/B3Value.h:
* Source/JavaScriptCore/b3/B3ValueInlines.h:
* Source/JavaScriptCore/b3/B3ValueKey.cpp:
(JSC::B3::ValueKey::materialize const):
* Source/JavaScriptCore/b3/air/AirOpcode.opcodes:
* Source/JavaScriptCore/wasm/WasmAirIRGenerator64.cpp:
(JSC::Wasm::AirIRGenerator64::addSIMDRelaxedFMA):
* Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp:
(JSC::Wasm::B3IRGenerator::addSIMDRelaxedFMA):
* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
(JSC::Wasm::BBQJIT::addSIMDRelaxedFMA):
* Source/JavaScriptCore/wasm/WasmFunctionParser.h:
(JSC::Wasm::FunctionParser&lt;Context&gt;::simd):
* Source/JavaScriptCore/wasm/WasmSIMDOpcodes.h:
(JSC::dumpSIMDLaneOperation):
(JSC::isRelaxedSIMDOperation):

Canonical link: <a href="https://commits.webkit.org/265324@main">https://commits.webkit.org/265324@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/efac7fc2262c408f1a869ab854e9bb9c3b47bade

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/10392 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/10623 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/10896 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/12033 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/9977 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/12723 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/10575 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/12940 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/10550 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/11339 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/8751 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/12430 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/8643 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/9407 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/16677 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/8780 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/9734 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/9558 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/12812 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/9854 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/10001 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/8110 "2 failures") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/10521 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/9159 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/2735 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2539 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/13411 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/10809 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/9854 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/2670 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

justinmichaud commented on Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp:
```diff
@@ -554,6 +554,20 @@ class B3IRGenerator {
+        auto b3Op = B3::Oops;
+        if (op == SIMDLaneOperation::RelaxedMAdd)
+            b3Op = B3::VectorRelaxedMAdd;
+        else if (op == SIMDLaneOperation::RelaxedNMAdd)
```
Let's use the macro here

justinmichaud reviewed (approved):
r=me if EWS passes

justinmichaud reviewed (approved):
r=me

danlliu committed:
Implement relaxed SIMD madd and nmadd instructions
https://bugs.webkit.org/show_bug.cgi?id=258047
rdar://110736062

Reviewed by Justin Michaud.

Added implementation and tests for f32x4.relaxed_madd, f32x4.relaxed_nmadd, f64x2.relaxed_madd, f64x2.relaxed_nmadd.

* JSTests/wasm/stress/simd-const-relaxed-f32-madd.js: Added.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.export.string_appeared_here.param.sz.i32.result.f32.i32.const.9.drop.v128.const.f32x4.5.0.4.0.37.0.6.0.v128.const.f32x4.22.0.25.0.3.0.20.0.v128.const.f32x4.1.0.1.0.0.0.1.0.f32x4.relaxed_madd.return.f32x4.extract_lane.1.func.export.string_appeared_here.param.sz.i32.result.f32.i32.const.9.drop.v128.const.f32x4.5.0.4.0.37.0.6.0.v128.const.f32x4.22.0.25.0.3.0.20.0.v128.const.f32x4.1.0.1.0.0.0.1.0.f32x4.relaxed_nmadd.return.f32x4.extract_lane.3.async test):
* JSTests/wasm/stress/simd-const-relaxed-f64-madd.js: Added.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.export.string_appeared_here.param.sz.i32.result.f64.i32.const.9.drop.v128.const.f64x2.28.0.24.0.v128.const.f64x2.10.0.20.0.v128.const.f64x2.1.0.2.0.f64x2.relaxed_madd.return.f64x2.extract_lane.0.func.export.string_appeared_here.param.sz.i32.result.f64.i32.const.9.drop.v128.const.f64x2.28.0.24.0.v128.const.f64x2.10.0.20.0.v128.const.f64x2.1.0.2.0.f64x2.relaxed_nmadd.return.f64x2.extract_lane.1.async test):
* Source/JavaScriptCore/assembler/ARM64Assembler.h:
* Source/JavaScriptCore/assembler/MacroAssemblerARM64.h:
(JSC::MacroAssemblerARM64::vectorFusedMulAdd):
(JSC::MacroAssemblerARM64::vectorFusedNegMulAdd):
* Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
(JSC::MacroAssemblerX86_64::vectorMulAdd):
* Source/JavaScriptCore/b3/B3LowerToAir.cpp:
* Source/JavaScriptCore/b3/B3Opcode.cpp:
(WTF::printInternal):
* Source/JavaScriptCore/b3/B3Opcode.h:
* Source/JavaScriptCore/b3/B3SIMDValue.h:
* Source/JavaScriptCore/b3/B3Validate.cpp:
* Source/JavaScriptCore/b3/B3Value.cpp:
(JSC::B3::Value::effects const):
(JSC::B3::Value::key const):
* Source/JavaScriptCore/b3/B3Value.h:
* Source/JavaScriptCore/b3/B3ValueInlines.h:
* Source/JavaScriptCore/b3/B3ValueKey.cpp:
(JSC::B3::ValueKey::materialize const):
* Source/JavaScriptCore/b3/air/AirOpcode.opcodes:
* Source/JavaScriptCore/wasm/WasmAirIRGenerator64.cpp:
(JSC::Wasm::AirIRGenerator64::addSIMDRelaxedFMA):
* Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp:
(JSC::Wasm::B3IRGenerator::addSIMDRelaxedFMA):
* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
(JSC::Wasm::BBQJIT::addSIMDRelaxedFMA):
* Source/JavaScriptCore/wasm/WasmFunctionParser.h:
(JSC::Wasm::FunctionParser<Context>::simd):
* Source/JavaScriptCore/wasm/WasmSIMDOpcodes.h:
(JSC::dumpSIMDLaneOperation):
(JSC::isRelaxedSIMDOperation):

Canonical link: https://commits.webkit.org/265324@main
